lib/Conversion/TosaToXTenNN.cpp

```diff
@@ -0,0 +1,225 @@
+/// and dequantize operations.
+class CastsToQDQOps : public OpRewritePattern<tosa::CastOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::CastOp castOp,
+                                PatternRewriter &rewriter) const override {
+    auto quantizeOp =
+        dyn_cast_or_null<tosa::CastOp>(castOp->getOperand(0).getDefiningOp());
+    if (!quantizeOp) {
+      return rewriter.notifyMatchFailure(castOp->getOperand(0).getLoc(),
+                                         "expected cast -> cast pattern.");
+    }
+
+    // Dequantize is from i8 -> f32 here we need to check for that
+    if (!isI8ToF32Cast(castOp)) {
+      return rewriter.notifyMatchFailure(
+          castOp->getLoc(), "dequantize operation input and output are not i8 "
+                            "and f32 respectively.");
+    }
+
+    // Quantize is from f32 -> i8 here we need to check for that
+    if (!isF32ToI8Cast(quantizeOp)) {
+      return rewriter.notifyMatchFailure(castOp->getLoc(),
+                                         "quantize operation input and output "
+                                         "are not f32 and i8 respectively.");
+    }
+
+    auto outputTensorType =
+        cast<TensorType>(quantizeOp->getOperand(0).getType());
+    Type newOutputType = outputTensorType.cloneWith(
+        {}, IntegerType::get(rewriter.getContext(), 8,
+                             IntegerType::SignednessSemantics::Signed));
+
+    auto newQuantizeOp = rewriter.create<amd::xten_nn::QuantizeOp>(
+        quantizeOp->getLoc(), newOutputType, quantizeOp->getOperand(0),
+        /*shift*/ (int32_t)0);
+    rewriter.replaceOpWithNewOp<amd::xten_nn::DequantizeOp>(
+        castOp, outputTensorType, newQuantizeOp->getResult(0),
+        /*shift*/ (int32_t)0);
+    rewriter.eraseOp(quantizeOp);
+
+    return success();
+  }
+};
+
+/// Folds multiplications surrounding a QDQ pair into the operations if the
+/// constants on the multiplications are power-of-two values and are equal.
+/// We assume the first multiplication represents the quantize meaning 1/scale
+/// and the second multiplication the dequantize or simply scale.
+class FoldMulsToQDQOps : public OpRewritePattern<tosa::MulOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::MulOp mulOp,
+                                PatternRewriter &rewriter) const override {
+
+    // The multiplication should have only a single constant value
+    APFloat dequantizeScaleFactor(0.0);
+    if (!m_ConstantFloat(&dequantizeScaleFactor)
+             .match(mulOp.getOperand(1).getDefiningOp())) {
+      return rewriter.notifyMatchFailure(mulOp.getOperand(1).getLoc(),
+                                         "expected to be a constant.");
+    }
+
+    // We want to make sure we have QDQ operations surrounded by MULs.
+    Operation *mulOperandOp = mulOp->getOperand(0).getDefiningOp();
```

Comment:
I think you can combine this with line `auto dequantizeOp = cast<amd::xten_nn::DequantizeOp>(mulOperandOp);` below into `auto dequantizeOp = mulOp->getOperand(0).getDefiningOp<amd::xten_nn::DequantizeOp>();`